### ISSUE TITLE
Question regarding reproducing the results on Cora

### ISSUE THREAD

**Issue by felixgwu** [lrjconan/LanczosNetwork#1]
Hi @lrjconan,

Thank you for sharing this awesome code!
LanczosNet works very well on QM8, and I am working on reproducing your results on Cora dataset.
With the hyperparameters given in the appendix of the paper, I can get 80.3 accuracy with one seed (slightly higher than the reported mean in the paper). However, it seems that LanczosNet has a very high variance and often underfits (both training and test accuracy are low).

There should be some parts in the paper that I misunderstood.
Would it be possible for you to read my code and help me figure them out?
Here is the code ([lanczos_net_cora.py](https://gist.github.com/felixgwu/14636d7afce34f0bec78e96839b60889)) of LanczosNet for Cora that I modified from the QM8 model.

> LanczosNet, we set the short and long diffusion scales to {1, 2, 5, 7} and {10, 20, 30} respectively.
The hidden dimension is 64 and dropout is 0.5. Lanczos step is 20. 1-layer MLP with 128 hidden
units and ReLU nonlinearity is used as the spectral filter.

Following this line, I use `nfeat=1433, nhid=64, nclass=7, dropout=0.5, num_layer=2, num_eig_vec=20, spectral_filter_kind='MLP', short_diffusion_dist=[1, 2, 5, 7], long_diffusion_dist=[10, 20, 30]`.
The nfeat and nclass are determined by the Cora dataset.

When I printed the model, I got
```
LanczosNet(
  (filter): ModuleList(
    (0): Linear(in_features=11464, out_features=64, bias=True)
    (1): Linear(in_features=512, out_features=7, bias=True)
  )
  (spectral_filter): ModuleList(
    (0): Sequential(
      (0): Linear(in_features=3, out_features=128, bias=True)
      (1): ReLU()
      (2): Linear(in_features=128, out_features=3, bias=True)
    )
    (1): Sequential(
      (0): Linear(in_features=3, out_features=128, bias=True)
      (1): ReLU()
      (2): Linear(in_features=128, out_features=3, bias=True)
    )
  )
)
```

The L, D, V are pre-computed using the `get_graph_laplacian_eigs` function in your 
[data_helper.py](https://github.com/lrjconan/LanczosNetwork/blob/master/utils/data_helper.py#L169-L258) script.

I use the training code from [https://github.com/tkipf/pygcn](https://github.com/tkipf/pygcn).
Since there is only a single graph, the input has no batch dimension. I unsqueeze the inputs and squeeze the outputs.

Thanks,
Felix

**Comment by lrjconan** [comment]
Thanks for you interests in our work!

**Comment by lrjconan** [comment]
Thanks for your interests in our work!

Regarding to your implementation, I think it is import to make sure:
* When you call `get_graph_laplacian_eigs` function, the argument `graph_laplacian_type` is set to `L4` as it is better than `L1` empirically.
* How do you handle the shape of L, is it of shape `1 X N X N X 1`?
* One way to possibly reduce the variance of Lanczos is to run `get_graph_laplacian_eigs` with slightly large value of `k`, e.g., 40, and only use the first 20. 

Also, as an FYI, there are some [known issues](https://sites.google.com/site/feast2015/program/downloads/orlova2015graph.pdf?attredirects=0&d=1) with these small citation datasets as they are fairly easy to overfit. Performances are typically very sensitive to hyperparameters like dropout. Therefore, they are not ideal and less conclusive for benchmarking.

Let me know if this helps to solve your problem.

**Comment by lrjconan** [comment]
Thanks for your interests in our work!

Regarding to your implementation, I think it is import to make sure:
  * When you call `get_graph_laplacian_eigs` function, the argument `graph_laplacian_type` is set to `L4` as it is better than `L1` empirically.
  * How do you handle the shape of `L`, is it of shape `1 X N X N X 1`?
  * One way to possibly reduce the variance of Lanczos is to run `get_graph_laplacian_eigs` with slightly large value of `k`, e.g., 40, and only use the first 20.

Also, as an FYI, there are some [known issues](https://arxiv.org/pdf/1811.05868.pdf) with these small citation datasets as they are fairly easy to overfit. Performances are typically very sensitive to hyperparameters like dropout. Therefore, they are not ideal and less conclusive for benchmarking.

Let me know if this helps to solve your problem.

**Comment by felixgwu** [comment]
Thank you for your instant reply!
- Yes, I use `L4`. I use this command to get the Laplacian. Hopefully, it's correct. `D, V, adj = get_graph_laplacian_eigs(adj, k=20, graph_laplacian_type='L4', use_eigen_decomp=True, is_sym=True)`
- The L is of shape `1 x N x N`. Since there is only one type of edges, I just drop the last dimension and use do `state = torch.bmm(L, state)` in my code.
- I'll try larger values. Thanks!

**Comment by felixgwu** [comment]
I realized that I should use `use_eigen_decomp=False` when decomposing the Laplacian.
Unfortunately, using a larger `k` doesn't stabilize the model. I tried up to 200.
AdaLanczosNet also has the same problem.
However, I don't think this is a big deal as long as people select models based on the validation score.
Additionally, in my experiments, I do find AdaLanczosNet works even better than LanczosNet.
Thank you for sharing this great work!

**Comment by lrjconan** [comment]
* `use_eigen_decomp=False` will use Lanczos, otherwise normal eigen solver will be used. Sorry for not well documented.
* `larger k doesn't stabilize the model`, this is good to know. The variance of the Lanczos algorithm is an interesting phenomenon. I will look into it once I got time.
* Interesting to know `AdaLanczosNet` works better, thanks for the information.
* I update the [known issues](https://arxiv.org/pdf/1811.05868.pdf) of citation networks with the correct paper link in case you are interested.
